codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/PresignURLAutoFill.java

```diff
@@ -71,6 +66,122 @@
+            GoWriter writer,
+            SymbolProvider symbolprovider,
+            OperationShape operation,
+            StructureShape input
+    ) {
+        Symbol operationSymbol = symbolprovider.toSymbol(operation);
+        Symbol inputSymbol = symbolprovider.toSymbol(input);
+
+        Symbol removeMiddleware = SymbolUtils.createValueSymbolBuilder("RemoveMiddleware",
+                AwsCustomGoDependency.PRESIGNEDURL_CUSTOMIZATION).build();
+
+        writer.openBlock(
+                "func $L(ctx context.Context, client interface{}, region string, params interface{}) "
+                        + "(req *v4.PresignedHTTPRequest, err error) {", "}",
+                presignFuncName(operationSymbol.getName(), false),
+                () -> {
+                    writer.addUseImports(SmithyGoDependency.FMT);
+                    // check input
+                    writer.write("input, ok := params.($P)", inputSymbol);
+                    writer.openBlock("if !ok {", "}", () -> {
+                        writer.write("return req, fmt.Errorf(\"expect $P type, got %T\", params)", inputSymbol);
+                    });
+
+                    // check client
+                    writer.write("c, ok := client.(*PresignClient)");
+                    writer.openBlock("if !ok {", "}", () -> {
+                        writer.write("return req, fmt.Errorf(\"expect *PresignClient type, got %T\", client)");
+                    });
+
+                    // generate client options
+                    writer.openBlock("optFn := func(o *Options) {", "}", () -> {
+                        writer.write("o.Region = region");
+                        writer.write("o.APIOptions = append(o.APIOptions, $T)", removeMiddleware);
+                    });
+
+                    // getPresignAPIOptions
+                    writer.write("presignOptFn := WithPresignClientFromClientOptions(optFn)");
+
+                    // call the exported function
+                    writer.write("return c.$L(ctx, input, presignOptFn)",
+                            presignFuncName(operationSymbol.getName(), true));
+                });
+    }
+
+    private static String addPresignMiddlewareFuncName(String operationName) {
+        return String.format("add%sPresignURLMiddleware", operationName);
+    }
+
+    private static String getterFuncName(String operationName, String memberName) {
+        return String.format("get%s%s", operationName, memberName);
+    }
+
+    private static String setterFuncName(String operationName, String memberName) {
+        return String.format("set%s%s", operationName, memberName);
+    }
+
+    private static String copyInputFuncName(String inputName) {
+        return String.format("copy%sForPresign", inputName);
+    }
+
+    private static String presignFuncName(String operationName, boolean exported) {
+        return exported ? String.format("Presign%s", operationName) : String.format("presign%s", operationName);
+    }
```

Comment:
These probably don't need to be exported, since they should only be used directly by the customization middleware.